eksrollup/lib/k8s.py

```diff
@@ -152,6 +152,141 @@ def taint_node(node_name):
         logger.info("Exception when calling CoreV1Api->patch_node: {}".format(e))
 
 
+def deployment_restart(node_name):
+    """
+    Availability is not guaranteed for a single replica deployment.
+    PDB only works with more than 1 replica. Node draining without
+    PDB is synchronous, meaning new pods are started at the same
+    time old pods are being terminated. A deployment restart will
+    ensure there is always one pod available to the service. If
+    enabled, this function will only perform a rolling restart on
+    deployments with single replicas.
+
+    This can raise an Kubernetes API exception, we're happy to let it bubble up
+    to the main caller and for the logic to terminate.
+    """
+
+    # Get the applicable deployments, this will return a list of tuples or raise an 
+    # exception that we can let bubble up to the main caller
+    deployments = find_applicable_deployments(node_name)
+
+    # Will track the sum of all return codes
+    total_return_codes = 0
+
+    with ThreadPoolExecutor(max_workers=10) as executor:
+        futures = []
+        for (deployment_name, deployment_namespace) in deployments:
+            future = executor.submit(rolling_restart, deployment_name, deployment_namespace)
+            futures.append(future)
+
+        # Attach threads
+        for thr in as_completed(futures):
+            returncode = thr.result()
+            total_return_codes += returncode
+
+        logger.info("Deployment restart function returned %s errors", total_return_codes)
```

Comment:
If the node has 0 deployment it needs to restart do we still want to log "Deployment restart function returned 0 errors"?

Reply:
I addressed this, thanks Nik!